rust/bridge/node/futures/src/exception.rs

```diff
@@ -0,0 +1,93 @@
+//
+// Copyright 2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use neon::prelude::*;
+
+/// See [PersistentException]
+enum PersistentExceptionImpl {
+    Object(Root<JsObject>),
+    String(String),
+}
+
+/// Persists a JavaScript exception across call contexts for later re-throwing
+/// (or promise rejection).
+///
+/// Like [neon::handle::Root][root], but falls back to preserving a string representation
+/// if the wrapped value is not a JavaScript Object.
+///
+/// PersistentException *must* be consumed explicitly using [drop][PersistentException::drop]
+/// or [into_inner][PersistentException::into_inner], like Root.
+/// This is because it must be unregistered with the JavaScript garbage collector.
+///
+/// [root]: https://docs.rs/neon/0.7.0-napi.3/neon/handle/struct.Root.html
+pub struct PersistentException {
+    wrapped: PersistentExceptionImpl,
```

Comment:
The wrapping is to keep clients from matching on the cases of the enum, which (as far as I know?) Rust doesn't allow you to prevent otherwise. The multiple representations are an implementation detail.